README.md
# Test User eXperience Harness
Build a stand-alone test harness with dynamic data to quickly prototype and isolate your HTML/CSS or examine your JSON document.

[![npm version](https://badge.fury.io/js/tuxharness.svg)](http://badge.fury.io/js/tuxharness)
[![Dependencies Status](https://david-dm.org/danactive/tuxharness.svg)](https://david-dm.org/danactive/tuxharness)
[![DevDependencies Status](https://david-dm.org/danactive/tuxharness/dev-status.svg)](https://david-dm.org/danactive/tuxharness#info=devDependencies)
[![MIT Licensed](http://img.shields.io/badge/license-MIT-blue.svg?style=flat-square)](http://opensource.org/licenses/MIT)

## Installation steps to setup and configure
1. `npm install tuxharness --save-dev`
1. Define a **recipe** file -- similar to *Gruntfile.js* or *gulpfile.js*
	* Mac/Unix copy file `cp -nv "node_modules/tuxharness/tuxfile.js" "tuxfile.js"`
	* Windows copy file `copy "node_modules/tuxharness/tuxfile.js" "tuxfile.js"`
	* More example recipes [starter sample tuxharness recipe](test/starter.js) or [full sample](test/sample.js)
1. Run the harness server
	* `node ./node_modules/tuxharness/index.js`
	* or add a *script* reference to your package.json like `"harness": "node ./node_modules/tuxharness/index.js"` then execute with `npm run harness`

## Recipe details
### Register
* `port` localhost server port number (default is *4000*)
* `static` serve static assets (img, js, css, fonts...)
	* `route` is a virtual web path - served by express.js
	* `directory` is the physical path where the static assets are hosted
* `view`
	* `engines` array of template view engine name as defined in *npm*
	* `path` is the physical path to the views folder - relative from your project root

### Harness (*array*)
* `route` is a virtual web path - served by express.js
* `view` is the physical path where the view is hosted - relative to your view folder
* `data` (*object*) - passed into the view for transformation
* `data` (*string*) - service call to request JSON data
* `data` (*function(**callback**, **util**)*) - callback argument will async call to gather data passed into the view for transformation
	* *callback* type string/JSON: print to browser
	* *util* type object
		* *getIpsumText* Lorum Ipsum generator
		* *getJsonRoute* type function(string route): wraps route name with localhost IP address with port and JSON path

## Template View Engines
1. [dust.js](https://github.com/linkedin/dustjs) (sample in test folder) 
	* `npm install dustjs-linkedin --save`
1. [Jade](https://github.com/jadejs/jade) (sample in test folder) 
	* `npm install jade --save`
1. Many others supported by [Consolidate.js](https://github.com/tj/consolidate.js)
	1. install any additional engines in your project
	1. register your view engine in your project's JS file

## Commands
* `npm run harness` Single run server to serve the test harness environment
* `npm run dev` Respawn-on-save server to serve the test harness environment
* `npm run lint` Lint JS files
* `npm test` Unit test and coverage

## JSON browser extentions
* Firefox [JSONView](https://addons.mozilla.org/en-US/firefox/addon/jsonview/)
* Chrome [JSONView](https://chrome.google.com/webstore/detail/jsonview/chklaanhfefbnpoihckbnefhakgolnmc)
* Safari [safari-json-formatter](https://github.com/rfletcher/safari-json-formatter)

## Roadmap
1. Allow remote recipes
	* Views and data too
1. Host demo


CHANGELOG.md
#Changelog
## v2.0.5 2016-Dec-15
* Lint
	* Lint and tests pass
	* Convert peerDependency to devDependency

## v2.0.4 2016-Nov-05
* Engine
	* Support all Node.js platform engines 
	
## v2.0.3 2016-Nov-01
* Dependencies
	* Add greenkeeper for dependency management

## v2.0.2 2015-Jun-19
* Dependencies
	* Update *gulp-expect-file* to v0.0.7
	* Update *gulp-istanbul-file* to v0.10

## v2.0.1 2015-Jun-02
* Dependencies
	* Update *gulp* to v3.9

## v2.0.0 2015-May-30
* Breaking change
	* `tuxfile.js`
		* Register view engine structure changed. For example `"view":{"engines":["dust","jade"],"path":"test/views"}`
		* Set harness route view changed to `"view": "hello.dust"`
	* Removed tuxharness property from `package.json`
* Bug
	* Killed [bug #2](issues/2)
* Data
	* Add Lorem Ipsum generator as util.getIpsumText
* Dependencies
	* Update *consolidate* to v0.13
	* Update *gulp-istanbul* to v0.9
	* Update *jshint-stylish* to v2.0

## v1.7.1 2015-Apr-18
* Installation
	* Easy tuxfile copy command

## v1.7.0 2015-Apr-11
* Refactor
	* Use prototype rather than formatted JSON
	* Add 50% unit test coverage
* Dependencies
	* Update *consolidate*
	* Update *gulp-jshint*
	* Update *gulp-load-plugins*

## v1.6.3 2015-Mar-28
* Data
	* Add util.getJsonRoute when data is a function

## v1.6.2 2015-Mar-28
* Bug
	* Fix when viewing a route without a view defined

## v1.6.1 2015-Mar-27
* Setup
	* tuxfile.js is default recipe filename
* Dependencies
	* Update *gulp-load-plugins* to v0.9
	* Update *request* to v2.54

## v1.5.2 2015-Mar-17
* Dependency
	* Update *serve-index* to v1.6.3 to avoid XSS security advisory

## v1.5.1 2015-Mar-08
* Dependencies
	* Update *dustjs-linkedin* to v2.6
	* Update *gulp-develop-server* to v0.4
* Bug
	* [Bug #2](issues/2) added

## v1.5.0 2015-Feb-27
* Support SSL for HTTPS requests
* Using npm ignore to hide dev files

## v1.4.2 2015-Feb-24
* Engine
	* Verified on iojs v1.3.0

## v1.4.1 2015-Feb-24
* Dependencies
	* Update *express* to v4.12
	* Update *gulp-develop-server* to v0.3

## v1.4.0 2015-Feb-22
* Dependency
	* Drop *request* as dependency, use built-in http module
		* Reason: request footprint is 2.6Mb and currently fails on node.js v0.12
* Home
	* Drop dust view engine dependency for home page
	* Redirect missing routes to home with error message
* Debug
	* Add new route for debugging recipe `tuxharness-debug-recipe`

## v1.3.1 2015-Feb-17
* Dependency
	* Update *consolidate.js* to v0.11

## v1.3.0 2015-Feb-17
* Data
	* JSON route changed from json?route=route to route/json

## v1.2.0 2015-Feb-17
* Static
	* Route and directory allow more flexibility

## v1.1.1 2015-Feb-15
* Home
	* Use route stem
* Route supports query string and hash

## v1.1.0 2015-Feb-10
* Home
	* Examine JSON
		* via data functions
		* via data string
		* via data require
		* via data embedding
	* Truncate long JSON
* Add changelog	

## v1.0.1 2015-Feb-06
* Readme improved
* Port default is `3000`

## v1.0.0 2015-Feb-05
* Initial release


test/sample.js
var streetAddress = {
		"address": {
			"street": "123 Main St",
			"city": "Vancouver",
			"province": "BC"
		}
	};
module.exports = {
	"register": {
		"port": 3000,
		"static": {
			"route": "static",
			"directory": "test/static"
		},
		"view": {
			"engines": ["dust", "jade"],
			"path": "test/views"
		}
	},
	"harnesses": [
		{
			"route": "html",
			"view": "html.dust"
		},
		{
			"route": "helloDust",
			"data": streetAddress,
			"view": "hello.dust"
		},
		{
			"route": "helloJade",
			"data": {
				"title": "Jade title",
				"message": "Jade view engine"
			},
			"view": "hello.jade"
		},
		{
			"route": "facebookFunction",
			"data": function (cb) {
				try {
					var request = require('request');
					request('http://graph.facebook.com/github', {"json": true}, function (error, response, body) {
						if (!error && response.statusCode === 200) {
							cb(body); // return JSON to view after async call via callback
						} else {
							throw new ReferenceError("Cannot connect to Facebook due to error: " + error);
						}
					});
				} catch (e) {
					cb(e);
				}
			},
			"view": "api.dust"
		},
		{
			"route": "facebookString",
			"data": 'http://graph.facebook.com/github',
			"view": "api.dust"
		},
		{
			"route": "facebookStringSSL",
			"data": 'https://graph.facebook.com/github',
			"view": "api.dust"
		},
		{
			"route": "largeJson",
			"data": require(__dirname + "/data/canadian-sales-tax.json"),
			"view": "salesTax.jade"
		},
		{
			"route": "rawLargeJson",
			"data": require(__dirname + "/data/canadian-sales-tax.json")
		},
		{
			"route": "routeQuery?city=Vancouver",
			"view": "html.dust",
			"data": streetAddress
		},
		{
			"route": "routeHash#anchor",
			"view": "html.dust",
			"data": streetAddress
		},
		{
			"route": "utilGetJsonRoute",
			"data": function (cb, util) {
				cb(util.getJsonRoute("largeJson"));
			}
		},
		{
			"route": "ipsum",
			"view": "ipsum.dust",
			"data": function (cb, util) {
				cb({"ipsum": util.getIpsumText()});
			}
		}
	]
};


test/starter.js
module.exports = {
	"register": {
		"static": {
			"route": "static",
			"directory": "public"
		},
		"view": {
			"engines": ["dust"],
			"path": "views"
		}
	},
	"harnesses": [
		{
			"route": "hello",
			"data": {
				"message": "Hello Fruits",
				"fruits": ["Apple", "Banana", "Cherry", "Durrian"]
			}
		},
		{
			"route": "facebook",
			"data": "http://graph.facebook.com/github"
		}
	]
};

test/unit_tests.js
var expect = require("expect.js"),
	Harness = require("../index.js"),
	sample = require("./sample.js");

describe('Harness', function () {
	function testHarness(args) {
		var harness = new Harness(sample.harnesses.filter(function (harness) {
			return harness.route === args.route;
		})[0]);
		expect(harness.isHomeRouteLinkable()).to.be(args.isHomeRouteLinkable);
		expect(harness.getHomeRoute()).to.be("/"+args.route);
		expect(harness.getHomeView()).to.be(args.getHomeView);
		if (args.getHomeDatum) {
			expect(harness.getHomeDatum()).to.be(args.getHomeDatum);
		}
		expect(harness.getRoute()).to.be("/"+args.getRoute);
		expect(harness.getJsonRoute()).to.be("/"+args.getJsonRoute+"/json");
		if (args.getJsonDatum) {
			expect(harness.getJsonDatum()).to.eql(args.getJsonDatum);
		}
		args.done();
	}
	it('should define a new object', function () {
		var harness = new Harness();
		expect(harness).to.have.keys("route", "data", "view");
		expect(harness.route).to.be(null);
		expect(harness.data).to.be(null);
		expect(harness.view).to.be(null);
		expect(harness.isHomeRouteLinkable()).to.be(false);
		expect(harness.getHomeRoute()).to.be("N/A");
		expect(harness.getHomeView()).to.be("N/A");
		expect(harness.isHomeDatumLinkable()).to.be(false);
		expect(harness.getHomeDatum()).to.be("N/A");
		expect(harness.getRoute()).to.be(null);
		expect(harness.getJsonRoute()).to.be(null);
		expect(harness.getJsonDatum()).to.eql({"Route is missing data": "N/A"});
	});
	it('route: html', function (done) {
		var route = "html";
		testHarness({
			"done": done,
			"route": route,
			"isHomeRouteLinkable": true,
			"getHomeView": "html.dust",
			"isHomeDatumLinkable": false,
			"getHomeDatum": "N/A",
			"getRoute": route,
			"getJsonRoute": route,
			"getJsonDatum": {"Route is missing data": "N/A"}
		});
	});
	it('route: helloDust', function (done) {
		var route = "helloDust";
		testHarness({
			"done": done,
			"route": route,
			"isHomeRouteLinkable": true,
			"getHomeView": "hello.dust",
			"isHomeDatumLinkable": true,
			"getHomeDatum": '{"address":{"street":"123 Main St","city":"Vancouver","province":"BC"}}',
			"getRoute": route,
			"getJsonRoute": route,
			"getJsonDatum": {"address":{"street":"123 Main St","city":"Vancouver","province":"BC"}}
		});
	});
	it('route: helloJade', function (done) {
		var route = "helloJade";
		testHarness({
			"done": done,
			"route": route,
			"isHomeRouteLinkable": true,
			"getHomeView": "hello.jade",
			"isHomeDatumLinkable": true,
			"getHomeDatum": '{"title":"Jade title","message":"Jade view engine"}',
			"getRoute": route,
			"getJsonRoute": route,
			"getJsonDatum": {"title":"Jade title","message":"Jade view engine"}
		});
	});
	it('route: facebookFunction', function (done) {
		var route = "facebookFunction";
		testHarness({
			"done": done,
			"route": route,
			"isHomeRouteLinkable": true,
			"getHomeView": "api.dust",
			"isHomeDatumLinkable": true,
			"getHomeDatum": "Service call function",
			"getRoute": route,
			"getJsonRoute": route
		});
	});
	it('route: facebookString', function (done) {
		var route = "facebookString";
		testHarness({
			"done": done,
			"route": route,
			"isHomeRouteLinkable": true,
			"getHomeView": "api.dust",
			"isHomeDatumLinkable": true,
			"getHomeDatum": "http://graph.facebook.com/github",
			"getRoute": route,
			"getJsonRoute": route
		});
	});
	it('route: facebookStringSSL', function (done) {
		var route = "facebookStringSSL";
		testHarness({
			"done": done,
			"route": route,
			"isHomeRouteLinkable": true,
			"getHomeView": "api.dust",
			"isHomeDatumLinkable": true,
			"getHomeDatum": "https://graph.facebook.com/github",
			"getRoute": route,
			"getJsonRoute": route
		});
	});
	it('route: largeJson', function (done) {
		var route = "largeJson";
		testHarness({
			"done": done,
			"route": route,
			"isHomeRouteLinkable": true,
			"getHomeView": "salesTax.jade",
			"isHomeDatumLinkable": true,
			"getRoute": route,
			"getJsonRoute": route
		});
	});
	it('route: rawLargeJson', function (done) {
		var route = "rawLargeJson";
		testHarness({
			"done": done,
			"route": route,
			"isHomeRouteLinkable": false,
			"getHomeView": "N/A",
			"isHomeDatumLinkable": true,
			"getRoute": route,
			"getJsonRoute": route
		});
	});
	it('route: routeQuery', function (done) {
		var route = "routeQuery";
		testHarness({
			"done": done,
			"route": route+"?city=Vancouver",
			"isHomeRouteLinkable": true,
			"getHomeView": "html.dust",
			"isHomeDatumLinkable": true,
			"getHomeDatum": '{"address":{"street":"123 Main St","city":"Vancouver","province":"BC"}}',
			"getRoute": route,
			"getJsonRoute": route,
			"getJsonDatum": {"address":{"street":"123 Main St","city":"Vancouver","province":"BC"}}
		});
	});
	it('route: routeHash', function (done) {
		var route = "routeHash";
		testHarness({
			"done": done,
			"route": route+"#anchor",
			"isHomeRouteLinkable": true,
			"getHomeView": "html.dust",
			"isHomeDatumLinkable": true,
			"getHomeDatum": '{"address":{"street":"123 Main St","city":"Vancouver","province":"BC"}}',
			"getRoute": route,
			"getJsonRoute": route,
			"getJsonDatum": {"address":{"street":"123 Main St","city":"Vancouver","province":"BC"}}
		});
	});
	it('route: utilGetJsonRoute', function (done) {
		var route = "utilGetJsonRoute";
		testHarness({
			"done": done,
			"route": route,
			"isHomeRouteLinkable": false,
			"getHomeView": "N/A",
			"isHomeDatumLinkable": true,
			"getHomeDatum": "Service call function",
			"getRoute": route,
			"getJsonRoute": route
		});
	});
	it('route: utilGetIpsum', function (done) {
		var route = "ipsum";
		testHarness({
			"done": done,
			"route": route,
			"isHomeRouteLinkable": true,
			"getHomeView": "ipsum.dust",
			"isHomeDatumLinkable": true,
			"getHomeDatum": "Service call function",
			"getRoute": route,
			"getJsonRoute": route
		});
	});
});

gulpfile.js
/*global Object */

var gulp = require('gulp'),
	notifier = require('node-notifier'),
	paths = {
		"lint": [".jshintrc", "package.json", "gulpfile.js", "index.js", "test/**/*.js"],
		"test": ["test/unit_tests.js"]
	},
	pkg = require('./package.json'),
	plugins = require('gulp-load-plugins')({"camelize": true});

gulp.task('lint', function() {
	return gulp.src(paths.lint)
		.pipe(plugins.expectFile(paths.lint))
		.pipe(plugins.jshint())
		.pipe(plugins.jshint.reporter('jshint-stylish'));
});

gulp.task('develop', ["lint"], function() {
	var getConfigPort,
		getViewFolders,
		recipe;

	getConfigPort = function () {
		var errorMsg = "Missing recipe filename or incorrect path. Definition must be package.json tuxharness key.";
		if (pkg.tuxharness === undefined) {
			throw new ReferenceError(errorMsg);
		}
		recipe = require(pkg.tuxharness);
		if (recipe === undefined) {
			throw new ReferenceError(errorMsg);
		}
		return recipe.register.port;
	};

	getViewFolders = function () {
		var paths = [];
		Object.keys(recipe.register.view).forEach(function (viewName) {
			paths.push("./" + recipe.register.view[viewName].path + "/**/*");
		});

		return paths;
	};

	plugins.developServer.listen({ "PORT": getConfigPort(), "path": pkg.main });

	gulp.watch(paths.lint, function () {
		gulp.start("lint", function () {
			plugins.developServer.restart();
			notifier.notify({
				'title': 'tuxharness server',
				'message': 'Restarted'
			});
		});
	});

	return gulp.watch(getViewFolders(), function () {
		plugins.developServer.restart();
		notifier.notify({
			'title': 'tuxharness server',
			'message': 'Restarted'
		});
	});
});

gulp.task('test', function() {
	return gulp.src(paths.test)
		.pipe(plugins.expectFile(paths.test))
		.pipe(plugins.mocha({reporter: 'nyan'}));
});

gulp.task('testcoverage', ['lint'], function(cb) {
	gulp.src(pkg.main)
		.pipe(plugins.expectFile(pkg.main))
		.pipe(plugins.istanbul()) // Covering files
		.pipe(plugins.istanbul.hookRequire()) // Force `require` to return covered files
		.on('finish', function () {
			gulp.src(paths.test)
				.pipe(plugins.expectFile(paths.test))
				.pipe(plugins.mocha({reporter: 'nyan'}))
				.pipe(plugins.istanbul.writeReports({
					dir: './temp/coverage',
					reportOpts: { dir: './temp/coverage' }
				})) // Creating the reports after tests runned
				.on('end', cb);
		});
});


lib/server.js
/*global process, require*/
(function () {
    'use strict';
    var app,
        appRoot,
        express = require('express'),
        fs,
        path = require("path"),
        recipe,
        recipeFilename = process && process.argv && process.argv.length >=3 && process.argv[2],
        server,
        serverPort = 3000,
        listDirectory = require('serve-index'),
        staticPath,
        template = require('consolidate'),
        viewPaths = [];
    app = express();
    if (recipeFilename === undefined) {
        throw new ReferenceError("Missing recipe filename or incorrect path");
    }
    recipe = require(recipeFilename);

    // register view engines based on recipe
    Object.keys(recipe.register.view).forEach(function (key) {
        app.engine(key, template[key]);
        app.set('view engine', key);
        viewPaths.push(recipe.register.view[key].path);
    });
    app.set('views', viewPaths);

    // static assets
    if (recipe.register.static) {
        appRoot = require('app-root-path');
        staticPath = path.join(appRoot.path, recipe.register.static.directory);
        app.use(express.static(staticPath)); // read file content /tux_static/js/sample.js
        if (recipe.register.static.route === "/") {
            fs = require('fs');

            fs.exists(staticPath, function (exists) {
                if (!exists) {
                    throw new ReferenceError("Static path cannot be found: " + staticPath);
                }
                fs.readdir(staticPath, function (err, filenames) {
                    if (err || filenames === undefined || filenames.length === 0) {
                        throw new ReferenceError("Static path cannot be read: " + staticPath);
                    }
                    recipe.register.static.route = [];
                    filenames.forEach(function (filename) {
                        recipe.register.static.route.push(filename);
                        app.use('/' + filename, listDirectory(path.join(appRoot.path, recipe.register.static.directory, filename), {'icons': true})); // serve directory listing
                    });
                });
            });
        } else {
            recipe.register.static.directory += "/" + recipe.register.static.route;
            recipe.register.static.route = [recipe.register.static.route];
            app.use('/' + recipe.register.static.route, listDirectory(path.join(appRoot.path, recipe.register.static.directory), {'icons': true})); // serve directory listing
        }
    }

    // home route
    app.get('/', function (req, res) {
        res.render(path.join(__dirname, "../test/views/dust/home.dust"), recipe);
    });

    // define routes based on recipe
    recipe.harnesses.forEach(function (harness) {
        app.get('/' + harness.route, function (req, res) {
            res.render(harness.view.path, harness.data);
        });
    });

    // start server
    server = app.listen(serverPort, function () {
        var host = server.address().address,
            port = server.address().port;

        console.log('tuxharness app running at http://%s:%s', host, port);
    });
})();


index.js
/*global module, require*/
var server,
	getServerAddress = function () {
		var host = server.address().address,
			port = server.address().port;
		return "http://" + ((host === "0.0.0.0" || host === "::") ? "localhost" : host) + ":" + port;
	},
	Harness = function Harness(definition) {
		definition = definition || {};
		this.route = definition.route || null;
		this.data = definition.data || null;
		this.view = definition.view || null;
	};
Harness.prototype.isHomeRouteLinkable = function () {
	return !(this.route === null || this.view === null);
};
Harness.prototype.getHomeRoute = function () {
	return (this.route === null) ? "N/A" : "/"+this.route;
};
Harness.prototype.getHomeView = function () {
	return (this.view === null) ? "N/A" : this.view;
};
Harness.prototype.isHomeDatumLinkable = function () {
	return !(this.route === null || this.data === null);
};
Harness.prototype.getHomeDatum = function () {
	var datum = null,
		dataLength = 300;
	switch (typeof this.data) {
		case "string":
			return this.data;
		case "function":
			return "Service call function";
		default:
			if (this.data === null) {
				return "N/A";
			}
			datum = JSON.stringify(this.data);
			return (datum.length > dataLength) ? datum.substring(0,dataLength) + "..." : datum;
	}
};
Harness.prototype.getRoute = function () {
	return (typeof this.route === "string") ? "/"+this.route.split(/[\?\#]/)[0] : null; // left of hash or query string
};
Harness.prototype.getJsonRoute = function () {
	return (this.getRoute() === null) ? null : this.getRoute() + "/json";
};
Harness.prototype.getJsonDatum = function () {
	return this.data || {"Route is missing data": "N/A"};
};

(function () {
	'use strict';
	var app,
		appRoot = require('app-root-path'),
		express = require('express'),
		harnesses = [],
		path = require("path");

	function setViewEngine(viewRule) {
		var template = require('consolidate');
		// register view engines based on recipe to express
		viewRule.engines.forEach(function (key) {
			app.engine(key, template[key]);
			app.set('view engine', key);
		});
		app.set('views', viewRule.path);
	}

	function getHomeHtml(model) {
		var html = [];

		html.push("<html><head><title>tuxharness - harness routes</title><style>.error{color: red;}</style></head><body>");

		if (model.msg) {
			html.push('<p class="error">', model.msg, '</p>');
		}

		if (harnesses.length > 0) {
			html.push("<h2>Your Routes</h2>");
			Array.prototype.push.apply(html, harnesses.map(function (harness) {
				var datum = harness.getHomeDatum(),
					datumHtml = datum+" "+ ((harness.isHomeDatumLinkable()) ? '<a href="'+harness.getJsonRoute()+'">Examine JSON</a>' : ""),
					route = harness.getHomeRoute(),
					routeHtml = (harness.isHomeRouteLinkable()) ? '<a href="'+route+'">'+route+'</a>' : route;
				return "<ul><li>Route "+routeHtml+"</li><li>View "+harness.getHomeView()+"</li><li>Data "+datumHtml+"</li></ul>";
			}));
		} else {
			html.push('<p class="error">Missing your harness routes.</p>');
		}

		if (model.register && model.register.static) {
			html.push("<h2>Static Route</h2>");
			html.push("<ul>");
			html.push('<li><a href="', model.register.static.route, '">', model.register.static.route, '</a> route points to ', model.register.static.directory, ' file folder.</li>');
			html.push("</ul>");
		}
		html.push("<h2>Recipe</h2>");
		html.push("<ul>");
		html.push('<li><a href="/tuxharness-debug-recipe">/tuxharness-debug-recipe</a></li>');
		html.push("</ul>");
		html.push("</body></html>");
		return html.join("");
	}

	function setHomeRoute(_recipe) {
		// express
		app.get('/', function (req, res) {
			if (req.query && req.query.msg) {
				_recipe.msg = req.query.msg;
			}
			res.send(getHomeHtml(_recipe));
		});
	}

	function setStaticRoute(staticRule) {
		var listDirectory = require('serve-index'),
			staticPath;
		staticPath = path.join(appRoot.path, staticRule.directory);

		app.use('/' + staticRule.route, express.static(staticPath)); // view document
		app.use('/' + staticRule.route, listDirectory(staticPath, {'icons': true})); // serve directory listing
	}

	function setCatchAllRoute(_recipe) {
		// express
		app.use("/tuxharness-debug-recipe", function(req, res) {
			res.send(_recipe);
		});

		app.use("/*", function(req, res) {
			var msg = "Route not defined: " + req.baseUrl;
			res.redirect(302, "/?msg=" + encodeURIComponent(msg));
		});
	}

	function getJsonViaString(url, cb) {
		var httpRequest,
			isSSL = (url.substring(0, 5) === "https");
		httpRequest = (isSSL) ? require('https') : require('http');
		httpRequest.get(url, function(response) {
			var body = '';

			// explicitly treat incoming data as utf8 (avoids issues with multi-byte chars)
			response.setEncoding('utf8');

			if (response.statusCode === 200) {
				// Continuously update stream with data
				response.on('data', function(data) {
					body += data;
				});
				response.on('end', function() {
					try {
						cb(null, JSON.parse(body));
					} catch (e) {
						cb(new TypeError("Service call failed, unable to parse response as JSON: " + e.message));
					}
				});
			} else {
				cb(new URIError("Service call failed with HTTP status code: " + response.statusCode));
			}
		}).on('error', function(e) {
			cb(new URIError("Service call failed due to error:" + e.message));
		});
	}

	function setExpressRoutes(harnesses) {
		var util = {
			"getIpsumText": function (options) {
				var defaultOptions = {
						count: 5,                        // Number of words, sentences, or paragraphs to generate.
						units: 'paragraphs',            // Generate words, sentences, or paragraphs.
						sentenceLowerBound: 5,         // Minimum words per sentence.
						sentenceUpperBound: 15,        // Maximum words per sentence.
						paragraphLowerBound: 3,        // Minimum sentences per paragraph.
						paragraphUpperBound: 7,        // Maximum sentences per paragraph.
						format: 'html',
						random: Math.random           // A PRNG function. Uses Math.random by default
					},
					extend = require("xtend"),
					ipsum = require('lorem-ipsum'),
					output;

				options = extend(defaultOptions, options);
				output = ipsum(options);
				return output;
			},
			"getJsonRoute": function (route) {
				if (route === undefined) {
					throw new ReferenceError("Missing route parameter");
				}
				return getServerAddress() + "/" + route + "/json";
			}
		};
		harnesses.forEach(function (harness) {
			// HTML route
			app.get(harness.getRoute(), function (req, res) {
				if (harness.data === null && harness.view === null) {
					res.status(404).send({"error": "The tuxharness recipe for " + harness.route + " needs either a view or data defined."});
				} else if (harness.view === null) {
					res.redirect(harness.json.route);
				} else if (harness.data === null) {
					res.render(harness.view);
				} else {
					switch (typeof harness.data) {
						case "string":
							getJsonViaString(harness.data, function (err, result) {
								if (err) {
									res.status(500).send({"error": err});
								} else {
									res.render(harness.view, result);
								}
							});
							break;
						case "function":
							harness.data(function (result) {
								res.render(harness.view, result);
							}, util);
							break;
						case "object":
							res.render(harness.view, harness.data);
							break;
						default:
							res.status(500).send({"error": "View template data is an odd format"});
							break;
					}
				}
			});
			// json route
			app.get(harness.getJsonRoute(), function (req, res) {
				switch (typeof harness.getJsonDatum()) {
					case "string":
						getJsonViaString(harness.getJsonDatum(), function (err, result) {
							if (err) {
								res.status(500).json({"error": err});
							} else {
								res.json(result);
							}
						});
						break;
					case "function":
						harness.getJsonDatum()(function (result) {
							res.json(result);
						}, util);
						break;
					default:
						res.json(harness.getJsonDatum());
						break;
				}
			});
		});
	}

	function init() {
		var recipe,
			serverPort;
		app = express();

		// load files
		try {
			recipe = require(path.join(appRoot.path, "tuxfile.js"));
		} catch (e) {
			throw new ReferenceError("Missing tuxfile.js recipe.");
		}

		serverPort = recipe.register && recipe.register.port || 4000;
		if (recipe.harnesses) {
			harnesses = recipe.harnesses.map(function (harness) {
				return new Harness(harness);
			});

			setExpressRoutes(harnesses);
		}
		if (recipe.register && recipe.register.view) {
			setViewEngine(recipe.register.view);
		}
		if (recipe.register && recipe.register.static) {
			setStaticRoute(recipe.register.static);
		}
		setHomeRoute(recipe);
		setCatchAllRoute(recipe);

		// start server
		server = app.listen(serverPort, function () {
			console.log('tuxharness app running at %s', getServerAddress());
		});
	}

	init();
})();

module.exports = Harness;
